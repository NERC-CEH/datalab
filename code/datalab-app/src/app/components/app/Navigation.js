import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import AppBar from 'material-ui/AppBar';
import Drawer from 'material-ui/Drawer';
import RaisedButton from 'material-ui/RaisedButton';
import NavMenu from './NavMenu';
import menuActions from '../../actions/menuActions';
import authActions from '../../actions/authActions';
import auth from '../../auth/auth';

class Navigation extends Component {
  constructor(props, context) {
    super(props, context);
    this.isUserLoggedIn = this.isUserLoggedIn.bind(this);
    this.userLoginLogout = this.userLoginLogout.bind(this);
  }

  componentWillMount() {
    const currentSession = auth.getCurrentSession();
    if (currentSession) {
      this.props.actions.userLogsIn(currentSession);
    }
  }

  isUserLoggedIn() {
    return this.props.user && auth.isAuthenticated(this.props.user);
  }

  userLoginLogout() {
    if (this.isUserLoggedIn()) {
      auth.logout();
    } else {
      auth.login();
    }
  }

  render() {
    return (
      <div>
        <AppBar
          title="Data Labs"
          onLeftIconButtonTouchTap={this.props.actions.showMenu}
          onRightIconButtonTouchTap={this.userLoginLogout}
          iconElementRight={
            <RaisedButton
              label={this.isUserLoggedIn() ? 'Logout' : 'Login'}
              secondary
            />}
        />
        <Drawer open={this.props.menu.isMenuOpen} docked={false} onRequestChange={this.props.actions.hideMenu}>
          <NavMenu hideMenu={this.props.actions.hideMenu} />
        </Drawer>
      </div>
    );
  }
}

function mapStateToProps({ menu, authentication: { user } }) {
  return {
    menu,
    user,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...authActions,
      ...menuActions,
    }, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Navigation);
