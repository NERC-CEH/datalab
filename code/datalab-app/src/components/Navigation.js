import React, { Component } from 'react';
import AppBar from 'material-ui/AppBar';
import Drawer from 'material-ui/Drawer';
import MenuItem from 'material-ui/MenuItem';

class Navigation extends Component {
  constructor(props) {
    super(props);
    this.state = { open: false };
    this.toggleDraw = this.toggleDraw.bind(this);
  }

  toggleDraw() {
    this.setState({ open: !this.state.open });
  }

  render() {
    return (
      <div>
        <AppBar title="Data Labs" onLeftIconButtonTouchTap={this.toggleDraw} />
        <Drawer open={this.state.open} docked={false} onRequestChange={this.toggleDraw}>
          <MenuItem>Home</MenuItem>
          <MenuItem>About</MenuItem>
        </Drawer>
      </div>
      );
  }
}

export default Navigation;
