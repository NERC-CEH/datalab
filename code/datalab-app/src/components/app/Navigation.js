import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import AppBar from 'material-ui/AppBar';
import Drawer from 'material-ui/Drawer';
import NavMenu from './NavMenu';
import menuActions from '../../actions/menuActions';

const Navigation = ({ isMenuOpen, actions: { showMenu, hideMenu } }) => (
  <div>
    <AppBar title="Data Labs" onLeftIconButtonTouchTap={showMenu} />
    <Drawer open={isMenuOpen} docked={false} onRequestChange={hideMenu}>
      <NavMenu hideMenu={hideMenu} />
    </Drawer>
  </div>
);

Navigation.propTypes = {
  isMenuOpen: PropTypes.bool.isRequired,
  actions: PropTypes.shape({
    showMenu: PropTypes.func.isRequired,
    hideMenu: PropTypes.func.isRequired,
  }).isRequired,
};

function mapStateToProps(state) {
  return state.menu;
}

function mapDispatchToProps(dispatch) {
  return { actions: bindActionCreators({ ...menuActions }, dispatch) };
}

export default connect(mapStateToProps, mapDispatchToProps)(Navigation);
