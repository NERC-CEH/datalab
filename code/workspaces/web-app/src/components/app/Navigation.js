import React from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { withStyles, useTheme } from '@material-ui/core/styles';
import TopBar from './TopBar';
import MessageBanner from './MessageBanner';

const styles = theme => ({
  container: {
    backgroundColor: theme.palette.backgroundColor,
    width: '100%',
    paddingTop: theme.shape.topBarHeight,
  },
  appFrame: {
    backgroundColor: theme.palette.backgroundColor,
    width: '100%',
    height: '100%',
    zIndex: 1,
  },
  pageContainer: {
    backgroundColor: theme.palette.backgroundColor,
    overflow: 'auto',
    flexDirection: 'column',
    width: '100%',
    display: 'flex',
  },
});

const Navigation = ({ classes, children, identity, userPermissions }) => {
  const messageCount = useSelector(s => s.messages.value).length;
  const theme = useTheme();
  const messagesHeight = messageCount * theme.shape.messageHeight;
  const heightDiff = theme.shape.topBarHeight + messagesHeight;

  return (
    <div className={classes.container} style={{ height: `calc(100vh - ${heightDiff}px)` }}>
      <div className={classes.appFrame}>
        <TopBar identity={identity} userPermissions={userPermissions} />
        <main className={classes.pageContainer} style={{ height: `calc(100% + ${messagesHeight}px)` }}>
          <MessageBanner/>
          {children}
        </main>
      </div>
    </div>
  );
};

Navigation.propTypes = {
  classes: PropTypes.object.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
  identity: PropTypes.object.isRequired,
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
  projectKey: PropTypes.string,
};

export default withStyles(styles)(Navigation);
