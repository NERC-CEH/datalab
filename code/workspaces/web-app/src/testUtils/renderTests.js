import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import theme from '../theme';

const AllTheProviders = ({ children }) => (
    <MuiThemeProvider theme={theme}>
        {children}
    </MuiThemeProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
