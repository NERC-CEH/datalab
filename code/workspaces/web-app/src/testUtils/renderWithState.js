import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import promise from 'redux-promise-middleware';
import {
  PROMISE_TYPE_PENDING,
  PROMISE_TYPE_SUCCESS,
  PROMISE_TYPE_FAILURE,
} from '../actions/actionTypes';

const promiseTypeSuffixes = [PROMISE_TYPE_PENDING, PROMISE_TYPE_SUCCESS, PROMISE_TYPE_FAILURE];

const TEST_USER = 'test-user';

export const buildDefaultTestState = () => ({
  authentication: {
    identity: {
      sub: TEST_USER,
    },
    permissions: {
      fetching: { inProgress: false },
      value: [],
    },
  },
  projectUsers: {
    fetching: { inProgress: false },
    value: [
      { userId: TEST_USER, role: 'admin' },
    ],
  },
});

export const renderWithState = (
  initialState,
  ComponentToRender,
  props = {},
  renderOptions = {},
) => {
  const mockStore = configureStore([promise({ promiseTypeSuffixes })]);
  const store = mockStore(initialState);

  const wrapper = render(
    <Provider store={store}>
      <ComponentToRender {...props} />
    </Provider>,
    renderOptions,
  );

  return {
    wrapper,
    store,
  };
};
