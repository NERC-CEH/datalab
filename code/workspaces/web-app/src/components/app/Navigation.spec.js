import React from 'react';
import { render } from '../../testUtils/renderTests';
import Navigation from './Navigation';

jest.mock('./MessageBanner', () => () => (<>MessageBanner Mock</>));
jest.mock('./TopBar', () => props => (<>Top Bar Mock {JSON.stringify(props)}</>));

describe('DescribeDatalabs', () => {
  it('renders correct snapshot', () => {
    const wrapper = render(
      <Navigation identity={{ expected: 'identity' }} userPermissions={['expectedPermission']} >
        <span>Content</span>
      </Navigation>,
    );

    expect(wrapper.container).toMatchSnapshot();
  });
});
