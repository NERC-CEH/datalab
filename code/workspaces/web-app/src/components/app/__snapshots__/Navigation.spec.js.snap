// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescribeDatalabs renders correct snapshot 1`] = `
<div
  className="Navigation-container-1"
>
  <div
    className="Navigation-appFrame-2"
  >
    <WithStyles(TopBar)
      identity={
        Object {
          "expected": "identity",
        }
      }
    />
    <main
      className="Navigation-pageContainer-3"
    >
      <span>
        Content
      </span>
    </main>
  </div>
</div>
`;
